packages/driver/cypress/integration/commands/assertions_spec.js

```diff
@@ -2927,4 +2927,23 @@ describe('src/cy/commands/assertions', () => {
       cy.get('.foo').should('not.exist')
     })
   })
+
+  context('implicit assertions', () => {
+    // https://github.com/cypress-io/cypress/issues/18549
+    it('only snapshots once when failing to find DOM elements', (done) => {
+      sinon.spy(cy, 'createSnapshot')
```

Comment:
you can work off of the event that cypress fires when it retries in order to count the number of retries that happened

- on the first retry, spy on cy.createSnapshot
- and then make sure that after it fails, there was only a single call to cy.createSnapshot
- and then it retried >1 times

that would tell you it took a snapshot on the final “error” state and not during the interstitial retries

its like... `cy.on('retry', ...)` or `cy.on('command:retry', ...)` I can't remember but there are other tests that use this event to count things.

that’ll give you a consistent and deterministic way to ensure that it both retried AND only created 1 snapshot and that test should obviously fail if we reverted the fix

Reply:
Reworked this into a more generic utility listening to cypress events, and added an event for snapshoting since we didn't already have one.

I also split the test into two pieces, one which precisely pinpoints the problem (a failing DOM assertion is taking three snapshots outside the context of retries, two of them because of `ensureElExistence`), and one which demonstrates it more dramatically (these calls because of `ensureElExistence` repeat with every retry, resulting in hundreds of snapshots).